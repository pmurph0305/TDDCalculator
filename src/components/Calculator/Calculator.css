.calculator-container {
  background-color: var(--display-background-color);
  height: var(--calculator-height);
  width: var(--calculator-width);
}
