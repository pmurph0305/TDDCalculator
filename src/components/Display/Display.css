.display-container {
  /* Flex */
  display: flex;
  align-items: center;

  /* Color */
  background: var(--display-background-color);

  /* Padding */
  padding: 0 5%;
  /* Dimensions */
  height: var(--display-height);
  width: var(--display-width);
}

.display-value {
  color: var(--display-text-color);
  font-size: var(--display-text-size);
  font-family: var(--main-font);
  font-weight: var(--display-text-weight);
  
  margin-left: auto;
  overflow: hidden;
}