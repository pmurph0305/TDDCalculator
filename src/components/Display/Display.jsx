import React from 'react';
import PropTypes from 'prop-types';

const Display = ({ displayedValue }) => (
<div className='display-container'>
    <p className='display-value'>{displayedValue}</p>
</div>)

Display.propTypes = { displayedValue: PropTypes.string.isRequired }

export default Display;