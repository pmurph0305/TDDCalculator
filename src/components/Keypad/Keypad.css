.keypad-container {
  /* Flex */
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  /* Dimensions */
  height: var(--keypad-height);
  width: var(--keypad-width);
  padding: 2%;
}

.numbers-container {
  /* Flex */
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  /* Dimensions */
  height: 80%;
  width: 75%;
}

.operators-container {
  /* Flex */
  display: flex;
  flex-direction: column;
  /* Dimensions */
  height: 80%;
  width: 25%;
}

.submit-container {
  height: 20%;
  width: 100%;
}

