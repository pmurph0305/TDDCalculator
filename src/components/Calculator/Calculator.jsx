import React from "react";

import Display from "../Display/Display";
import KeyPad from "../Keypad/Keypad";

import "./Calculator.css";

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // Value to be displayed with Display
      displayedValue: "0",
      // Numbers for Keys
      numbers: ["9", "8", "7", "6", "5", "4", "3", "2", "1", ".", "0", "CE"],
      // Operators for Keys
      operators: ["+", "-", "/", "*"],
      // Operator selected for math
      selectedOperator: "",
      // Stored calculated value for math
      storedValue: '0'
    };
  }

  updateDisplay = value => {
    let { displayedValue } = this.state;
    if (displayedValue)
      if (displayedValue === "0") {
        displayedValue = value;
      } else if (value === "CE") {
        if (displayedValue.length > 1) {
          displayedValue = displayedValue.slice(0, displayedValue.length - 1);
        } else {
          displayedValue = "0";
        }
      } else if (value === ".") {
        if (displayedValue.includes(".") === false) {
          displayedValue += ".";
        }
      } else {
        displayedValue += value;
      }
    this.setState({ displayedValue: displayedValue });
  };

  callOperator = () => {
    let { displayedValue, storedValue, selectedOperator } = this.state;
    displayedValue = parseInt(displayedValue);
    storedValue = parseInt(storedValue);

    if (selectedOperator === '+') {
      displayedValue = storedValue + displayedValue;
    } else if (selectedOperator === '-') {
      displayedValue = storedValue - displayedValue;
    } else if (selectedOperator === '*') {
      displayedValue = storedValue * displayedValue;
    } else if (selectedOperator === '/') {
      displayedValue = storedValue / displayedValue;
    }

    this.setState({ displayedValue: displayedValue.toString()})
  };

  setOperator = (operator) => {
    let { selectedOperator, displayedValue, storedValue } = this.state;
    if (selectedOperator === '') {
      storedValue = displayedValue;
      displayedValue = '0'
    }
    selectedOperator = operator;
    this.setState({ selectedOperator, displayedValue, storedValue })
  };

  render() {
    const { displayedValue, numbers, operators } = this.state;
    return (
      <div className="calculator-container">
        <Display displayedValue={displayedValue} />
        <KeyPad
          callOperator={this.callOperator}
          numbers={numbers}
          operators={operators}
          setOperator={this.setOperator}
          updateDisplay={this.updateDisplay}
        />
      </div>
    );
  }
}

export default Calculator;
