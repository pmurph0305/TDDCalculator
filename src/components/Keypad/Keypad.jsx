import React from "react";
import PropTypes from "prop-types";

import './Keypad.css'

const KeyPad = ({
  callOperator,
  numbers,
  operators,
  setOperator,
  updateDisplay
}) => {
  const numberKeys = numbers.map(num => <p key={num}>{num}</p>);
  const operatorKeys = operators.map(operator => (
    <p key={operator}>{operator}</p>
  ));

  return (
    <div className="keypad-container">
      <div className="numbers-container">{numberKeys}</div>
      <div className="operators-container">{operatorKeys}</div>
    </div>
  );
};

KeyPad.propTypes = {
  callOperator: PropTypes.func.isRequired,
  numbers: PropTypes.array.isRequired,
  operators: PropTypes.array.isRequired,
  setOperator: PropTypes.func.isRequired,
  updateDisplay: PropTypes.func.isRequired
};

export default KeyPad;
