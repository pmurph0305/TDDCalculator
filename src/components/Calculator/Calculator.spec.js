import React from "react";

import { shallow, mount } from "enzyme";

import Calculator from "./Calculator";
import Display from "../Display/Display";
import Keypad from '../Keypad/Keypad';

describe("Shallow Calculator tests", () => {
  let wrapper;

  beforeEach(() => (wrapper = shallow(<Calculator />)));

  it("Renders a div", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });

  it("Should render Display & Keypad Components", () => {
    expect(
      wrapper.containsAllMatchingElements([
        <Display displayedValue={wrapper.instance().state.displayedValue} />,
        <Keypad 
          callOperator={wrapper.instance().callOperator}
          numbers={wrapper.instance().state.numbers}
          operators={wrapper.instance().state.operators}
          setOperator={wrapper.instance().setOperator}
          updateDisplay={wrapper.instance().updateDisplay}
        />
      ])
    ).toEqual(true);
  });

  it('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

});

describe('Mounted Calculator Tests', () => {
  let wrapper;

  beforeEach(() => (wrapper = mount(<Calculator />)));

  it('Calls update display when a number key is clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateDisplay');
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find('.number-key').first().simulate('click');
    wrapper.find('.number-key').first().simulate('click');
    expect(spy).toHaveBeenCalledTimes(2);
  })

})
