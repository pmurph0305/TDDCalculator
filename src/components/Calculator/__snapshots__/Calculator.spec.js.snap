// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator tests Should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Calculator />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Display
          displayedValue="0"
/>,
        <KeyPad
          callOperator={undefined}
          numbers={Array []}
          operators={Array []}
          setOperator={undefined}
          updateDisplay={undefined}
/>,
      ],
      "className": "calculator-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "displayedValue": "0",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "callOperator": undefined,
          "numbers": Array [],
          "operators": Array [],
          "setOperator": undefined,
          "updateDisplay": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Display
            displayedValue="0"
/>,
          <KeyPad
            callOperator={undefined}
            numbers={Array []}
            operators={Array []}
            setOperator={undefined}
            updateDisplay={undefined}
/>,
        ],
        "className": "calculator-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "displayedValue": "0",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "callOperator": undefined,
            "numbers": Array [],
            "operators": Array [],
            "setOperator": undefined,
            "updateDisplay": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
